#### Viernes 16 de mayo de 2025

## Reglas de Asociación. Algoritmo Apriori. Canasta Básica

In [83]:
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules
from mlxtend.preprocessing import TransactionEncoder

In [84]:
# Transacciones

transacciones = [
    ['Pan', 'Leche'],
    ['Pan', 'Pañales', 'Cerveza'],
    ['Leche', 'Pañales'],
    ['Pan', 'Leche', 'Pañales', 'Huevos'],
    ['Huevos', 'Leche'],
    ['Pan', 'Huevos'],
    ['Pañales', 'Cerveza'],
    ['Pan', 'Leche', 'Huevos']
]

print("Dataset Original: \n")
for i, transaccion in enumerate(transacciones, 1):
    print(f"Transacción {i}: {transaccion}")

Dataset Original: 

Transacción 1: ['Pan', 'Leche']
Transacción 2: ['Pan', 'Pañales', 'Cerveza']
Transacción 3: ['Leche', 'Pañales']
Transacción 4: ['Pan', 'Leche', 'Pañales', 'Huevos']
Transacción 5: ['Huevos', 'Leche']
Transacción 6: ['Pan', 'Huevos']
Transacción 7: ['Pañales', 'Cerveza']
Transacción 8: ['Pan', 'Leche', 'Huevos']


### Preprocesamiento de Datos

In [ ]:
te = TransactionEncoder()

# Matriz booleana donde True indica la presencia del producto en la transacción
matriz = te.fit(transacciones).transform(transacciones)

df = pd.DataFrame(matriz, columns=te.columns_)

print("Dataset Transormado (One-Hot Encoding): \n")
print(df)

Dataset Transormado (One-Hot Encoding): 



Cerveza  Huevos  Leche    Pan  Pañales
0    False   False   True   True    False
1     True   False  False   True     True
2    False   False   True  False     True
3    False    True   True   True     True
4    False    True   True  False    False
5    False    True  False   True    False
6     True   False  False  False     True
7    False    True   True   True    False

### Algoritmo Apriori. Items Frecuentes

In [ ]:
items_frecuentes = apriori(df, min_support=0.3, use_colnames=True)

print("Items Frecuentes (Soporte >= 30%): \n")
print(items_frecuentes.sort_values('support', ascending=False))

Items Frecuentes (Soporte >= 30%): 



support         itemsets
1    0.625          (Leche)
2    0.625            (Pan)
0    0.500         (Huevos)
3    0.500        (Pañales)
4    0.375  (Huevos, Leche)
5    0.375    (Pan, Huevos)
6    0.375     (Pan, Leche)

### Reglas de Asociación

In [ ]:
# Reglas con confianza mínima del 60%
reglas = association_rules(items_frecuentes, metric="confidence", min_threshold=0.6)

#Filtramos reglas con lift > 1
reglas = reglas[reglas['lift'] > 1]

print("\nReglas de Asociación (Confianza >= 60%):\n")
print(reglas[['antecedents', 'consequents', 'support', 'confidence', 'lift']].sort_values('lift', ascending=False))


Reglas de Asociación (Confianza >= 60%):



antecedents consequents  support  confidence  lift
0    (Huevos)     (Leche)    0.375        0.75   1.2
1     (Leche)    (Huevos)    0.375        0.60   1.2
2       (Pan)    (Huevos)    0.375        0.60   1.2
3    (Huevos)       (Pan)    0.375        0.75   1.2

### Pregunta Clave

#### ¿Qué producto está más asociado a "Pan"? Usa el lift como criterio.

##### Pan aparece como antecedente y como consecuente en dos reglas, ambos tienen un lift de 1.2, entonces el producto más asociado es Huevos. Esto significa que comprar Pan aumenta la probabilidad de comprar Huevos en un 20% respecto a lo normal.